compaction/compaction_manager.cc

```diff
@@ -353,7 +353,7 @@ class compaction_manager::major_compaction_task : public compaction_manager::tas
         // candidates are sstables that aren't being operated on by other compaction types.
         // those are eligible for major compaction.
         auto* t = _compacting_table;
-        sstables::compaction_strategy cs = t->get_compaction_strategy();
+        auto& cs = t->get_compaction_strategy();
```

Comment:
Maybe we should delete the copy constructor?

Reply:
(and copy assignment operator)

Reply:
Done in v2

Reply:
I don't follow.

compaction_strategy is:

class compaction_strategy {
    ::shared_ptr<compaction_strategy_impl> _compaction_strategy_impl;

so copying strategy won't instantiate an actual copy of the state, but rather share the ownership of it.

Reply:
<blush>you're absolutely right</blush>
That said, I think that obfuscates the code that isn't WYSIWYG anymore.
I think it would be better to have a unique_ptr in class compaction_strategy and to keep an explicit shared / lw_shared pointer where needed so it would be clearer where we just get a shared pointer.

Reply:
Yes, one of the problems with using shared ptr underneath is that you break the copy semantics. copy with shared ptr means that we're getting a sort of reference to the original state, not an actual copy of it.